Translation of reference-react-dom (#13)
reactjs/pt-PT.reactjs.org: content/docs/reference-react-dom.md

permalink: docs/react-dom.html
---

Se usares o React através da tag `<script>`, estas APIs de alto nível estão disponíveis através do objecto global `ReactDOM`. Se usares ES6 através do npm, podes escrever `import ReactDOM from 'react-dom'`. Se usares ES5 através do npm, podes escrever `var ReactDOM = require('react-dom')`.

## Visão Geral {#overview}

> O pacote `react-dom` fornece métodos DOM específicos que podem user usados ao mais alto nível da tua aplicação como uma porta de salvação para saíres do modelo de React se precisares. A maioria dos teus componentes não deverão necessitar de usar este módulo.

Author: Podes editar a sugestão, e trocar "user" por "ser" sff?

> O pacote `react-dom` fornece métodos DOM específicos que podem user usados ao mais alto nível da tua aplicação como uma porta de salvação para saíres do modelo de React se precisares. A maioria dos teus componentes não deverão necessitar de usar este módulo.

Reviewer: Done!

Se o _callback_ opcional for fornecido, este irá ser executado após a renderização ou actualização do componente.

> Nota:
>
> `ReactDOM.render()` controla o conteúdo do nó (_node_) do _container_ que lhe é passado. Todos os elementos DOM que estão dentro do nó são substituídos na primeira invocação. As invocações seguintes usam um algoritmo de diferenciação do DOM (do próprio React) para actualizações eficientes.
>
> `ReactDOM.render()` não modifica o nó (_node_) do _container_ (apenas modifica os filhos do _container_). É portanto possível, inserir um componente num nó já existente no DOM sem reescrever os filhos existentes.
>
> `ReactDOM.render()` retorna a referência para a raíz da instância `ReactComponent`. No entanto, deverás evitar usá-la, pois em futuras versões de React, os componentes poderão ser renderizados assíncronamente em algums casos. Se necessitares usar uma referência para a raíz da instância `ReactComponent`, a melhor solução será anexar uma [referência callback (_callback ref_)](/docs/more-about-refs.html#the-ref-callback-attribute) no elemento raíz.
>

> > Usar `ReactDOM.render()` para "hidratar" um _container_ renderizado pelo servidor está descontinuado e será removido no React 17. Usa antes o método [`hydrate()`](#hydrate).

Reviewer 2: Este `[...] para "hidratar"` soa bastante estranho.

Author: Documentação em português é que soa estranho =P
Mas, alguma sugestão?

Reviewer: Acho que podiamos manter 'hidratar'. No meu entender o tal hidratar é preencher um objecto com dados.  Ou talvez fazer:

```suggestion
> Usar `ReactDOM.render()` para preencher com dados (_to hydrate_) um _container_ renderizado pelo servidor está descontinuado e será removido no React 17. Usa antes o método [`hydrate()`](#hydrate).
```

ReactDOM.hydrate(element, container[, callback])
```

O mesmo que [`render()`](#render), mas é usado para "hidratar" um _container_ cujos conteúdos HTML foram servidos por [`ReactDOMServer`](/docs/react-dom-server.html). O React irá tentar anexar manipuladores de eventos (_event handlers_) no código existente.

> O React espera que o conteúdo apresentado seja idêntico entre o servidor e o cliente. Pode corrigir diferenças de texto, mas deverás tratar divergências como _bugs_ e corrigi-las. Em modo de desenvolvimento, o React alerta sobre divergências durante o processo de "hidratação". Não há garantias que diferenças de atributos sejam corrigidas em caso de divergências. Isto é importante em termos de desempenho na maioria das aplicações, divergências são raras, e portanto, validar todo o código seria incomportávelmente dispendioso.

Reviewer: @jbarradas na última linha, "validar todo o código seria **incomportávelmente** dispendioso" não achas melhor trocar por "validar todo o código seria **exageradamente** dispendioso"?

cc: @telmogoncalves

Se um único atributo ou texto de um elemento for inevitavelmente diferente entre o servidor e o cliente (por exemplo, uma data e hora), o alerta poderá ser silenciado adicionando `suppressHydrationWarning={true}` ao elemento. Apenas funcionará no primeiro nível, e é suposto funcionar como uma escapatória. Não o uses excessivamente. A menos que se trate de texto, o React nem irá tentar corrigi-lo, portanto poderá permanecer intacto até uma próxima actualização.

> Se for necessário apresentar intencionalmente algo diferente entre o servidor e o cliente, poderás optar por uma renderização de dupla passagem (_two-pass rendering_). Componentes que apresentem algo diferente no cliente podem ler uma variável de estado como `this.state.isClient`, que poderá ser definida como `true` no `componentDidMount()`. Desta forma, a passagem inicial de apresentação, irá apresentar o mesmo conteúdo que o servidor, evitando divergências, mas uma passagem adicional irá acontecer sincronamente, logo a seguir à "hidratação". Nota que esta abordagem irá fazer como que os teus componentes sejam mais lentos, pois irão ser apresentados duas vezes, portanto, usa com precaução.

Reviewer 2: ```suggestion
Se for necessário apresentar intencionalmente algo diferente entre o servidor e o cliente, poderás optar por uma renderização de dupla passagem (_two-pass rendering_). Componentes que apresentem algo diferente no cliente podem ler uma variável de estado como `this.state.isClient`, que poderá ser definida como `true` no `componentDidMount()`. Desta forma, a passagem inicial de apresentação, irá apresentar o mesmo conteúdo que o servidor, evitando divergências, mas uma passagem adicional irá acontecer sincronamente, logo a seguir à "hidratação". Nota que esta abordagem irá fazer com que os teus componentes sejam mais lentos, pois irão ser apresentados duas vezes, portanto, usa com precaução.
```

> Sê consciente da experiência de utilizador em conexões lentas. O código Javascript pode ser carregado significantemente depois da apresentação inicial do HTML, portanto, se apresentares algo diferente na passagem do lado do cliente, a transição pode ser dissonante. No entanto, se bem executada, pode ser benéfico apresentar uma "_shell_" da aplicação no servidor, e apenas alguns _widgets_ extra no cliente. Para aprender como fazer isto sem ter problemas com as divergências no código, consulta a explicação no parágrafo anterior.

Reviewer: ```suggestion
Toma em consideração a experiência do utilizador em conexões lentas. O código JavaScript pode ser carregado significantemente depois da apresentação inicial do HTML, portanto, se apresentares algo diferente na passagem do lado do cliente, a transição pode ser dissonante. No entanto, se bem executada, pode ser benéfico apresentar uma "_shell_" da aplicação no servidor, e apenas alguns _widgets_ extras no cliente. Para aprender como fazer isto sem ter problemas com as divergências no código, consulta a explicação no parágrafo anterior.
```

ReactDOM.unmountComponentAtNode(container)
```

Remove um componente React montado do DOM e limpa o seu estado e manipuladores de eventos (_event handlers_). Se nenhum componente foi montado no _container_, a invocação deste método não tem qualquer efeito. Retorna `true` se o component foi desmontado e `false` se não houver nenhum componente para desmontar.

* * *

### `findDOMNode()` {#finddomnode}

> Nota:
>

> > `findDOMNode` é uma escapatória usada para aceder ao nó (_node_) do DOM subjacente. Na maioria dos casos, o uso desta escapatória é desaconselhado porque trespassa a abstração do componente. [Foi descontinuado em `StricMode`.](/docs/strict-mode.html#warning-about-deprecated-finddomnode-usage)

Reviewer: ```suggestion
> `findDOMNode` é uma escapatória usada para aceder ao nó (_node_) do DOM subjacente. Na maioria dos casos, o uso desta escapatória é desaconselhado porque trespassa a abstração do componente. [Foi descontinuado em `StrictMode`.](/docs/strict-mode.html#warning-about-deprecated-finddomnode-usage)
```

> > `findDOMNode` é uma porta de salvação usada para aceder ao nó (_node_) do DOM subjacente. Na maioria dos casos, o uso desta porta de salvação é desaconselhado porque trespassa a abstração do componente. [Foi deprecado em `StricMode`.](/docs/strict-mode.html#warning-about-deprecated-finddomnode-usage)

Reviewer: ```suggestion
> `findDOMNode` é uma escapatória usada para aceder ao nó (_node_) do DOM subjacente. Na maioria dos casos, o uso desta escapatória é desaconselhado porque trespassa a abstração do componente. [Foi descontinuado em `StricMode`.](/docs/strict-mode.html#warning-about-deprecated-finddomnode-usage)
```

```javascript
ReactDOM.findDOMNode(component)
```

> Se este componente foi montado no DOM, retorna o elemento DOM nativo correspondente. Este método é util para ler valores fora do DOM, como valores de um campo de um formulário e fazer medições ao DOM. **Na maioria dos casos, pode ser anexada uma referência (ref) no nó do DOM evitando o uso de `findDOMNode` por completo.**

Reviewer 2: ```suggestion
Se este componente foi montado no DOM, retorna o elemento DOM nativo correspondente. Este método é util para ler valores fora do DOM, como valores de um campo de um formulário e fazer medições ao DOM. **Na maioria dos casos, pode ser anexada uma referência (ref) no nó (_node_) do DOM evitando o uso de `findDOMNode` por completo.**
```

> Se este componente foi montado no DOM, retorna o elemento DOM nativo correspondente. Este método é util para ler valores fora do DOM, como valores de um campo de um formulário e realizar avaliações ao DOM. **Na maioria dos casos, pode ser anexada uma referência (ref) no nó do DOM evitando o uso de `findDOMNode` por completo.**

Reviewer: ```suggestion
Se este componente foi montado no DOM, retorna o elemento DOM nativo correspondente. Este método é util para ler valores fora do DOM, como valores de um campo de um formulário e fazer medições ao DOM. **Na maioria dos casos, pode ser anexada uma referência (ref) no nó do DOM evitando o uso de `findDOMNode` por completo.**
```

> Quando um componente retorna `null` ou `false`, `findDOMNode` retorna `null`. Se o componente retorna uma string, `findDOMNode` retorna um nó do DOM de texto contendo o valor dessa string. Desde o React 16, um componente pode retornar um fragmento com múltiplos filhos, e neste caso, `findDOMNode` irá retornar o nó do DOM correspondente ao primeiro filho não vazio.

Reviewer 2: ```suggestion
Quando um componente retorna `null` ou `false`, `findDOMNode` retorna `null`. Se o componente retorna uma string, `findDOMNode` retorna um nó (_node_) do DOM de texto contendo o valor dessa string. Desde o React 16, um componente pode retornar um fragmento com múltiplos filhos, e neste caso, `findDOMNode` irá retornar o nó (_node_) do DOM correspondente ao primeiro filho não vazio.
```

> Nota:
>

> > `findDOMNode` apenas funciona em componentes montados (isto é, componentes que tenham sido inseridos no DOM). Se tentares invocar este método num componente que ainda não tenha sido montado (por exemplo, invocando `findDOMNode` no `render()` de um componente que ainda não tenha sido criado) irá ser lançada uma excepção.

Reviewer: ```suggestion
> `findDOMNode` apenas funciona em componentes montados (isto é, componentes que tenham sido inseridos no DOM). Se tentares invocar este método num componente que ainda não tenha sido montado (por exemplo, invocando `findDOMNode` no `render()` de um componente que ainda não tenha sido criado) será lançada uma excepção.
```

ReactDOM.createPortal(child, container)
```

> Cria um portal. Portais fornecem uma forma de [renderizar filhos num nó do DOM que existe fora da hierárquia do DOM do componente](/docs/portals.html).

Reviewer 2: ```suggestion
Cria um portal. Portais fornecem uma forma de [renderizar filhos num nó (_node_) do DOM que existe fora da hierárquia do DOM do componente](/docs/portals.html).
```

> Cria um portal. Portais fornecem uma forma de [apresentar filhos num nó do DOM que existam fora da hierárquia do DOM do componente](/docs/portals.html).

Reviewer: ```suggestion
Cria um portal. Portais fornecem uma forma de [renderizar filhos num nó do DOM que existe fora da hierárquia do DOM do componente](/docs/portals.html).
```
